Replace aws sdk client

davidecarlier opened the pull request:
## Description

This update replaces the usage of aws-sdk in the drive-s3 package with @aws-sdk/client-s3. 

Fixes #16059 

## Type of Change

- [ ] Bugfix
- [x] Improvement
- [ ] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [ ] New / updated tests are included
- [x] All tests are passing locally
- [x] Performed a self-review of the submitted code

If adding a new feature:

- [ ] Documentation was added/updated. PR:

davidecarlier committed:
install @aws-sdk/client-s3

davidecarlier committed:
install @aws-sdk/client-s3

davidecarlier committed:
replace S3 with S3Client, fix methods copy and delete

davidecarlier committed:
replace S3 with S3Client, fix methods copy and delete

davidecarlier committed:
replace methods exists, getBuffer, getStat

davidecarlier committed:
replace methods exists, getBuffer, getStat

davidecarlier committed:
install @aws-sdk/s3-request-presigner and replace method getSignedUrl

davidecarlier committed:
install @aws-sdk/s3-request-presigner and replace method getSignedUrl

davidecarlier committed:
Merge branch 'directus:main' into replace-aws-sdk-client

davidecarlier committed:
Merge branch 'replace-aws-sdk-client' of https://github.com/davidecarlier/directus into replace-aws-sdk-client

davidecarlier committed:
wip

davidecarlier committed:
fix client initialization & getStream

davidecarlier committed:
fix put command

davidecarlier committed:
fix exists & flatList

davidecarlier committed:
fix geUrl

davidecarlier committed:
Merge commit '32a363b8038e29130bd1d9a084d360776bc2e150' into replace-aws-sdk-client

paescuj commented on packages/drive-s3/src/AmazonWebServicesS3Storage.ts:
```diff
@@ -147,14 +171,11 @@ export class AmazonWebServicesS3Storage extends Storage {
 
 		const params = { Key: location, Bucket: this.$bucket };
-
+		const result = await this.$driver.send(new GetObjectCommand(params));
+		const body = result.Body as NodeJS.ReadableStream;
```
This should probably go in to the `try...catch` block

davidecarlier replied:
Yep! fixed

davidecarlier committed:
fix getBuffer error handling

rijkvanzanten committed:
Merge branch 'main' into replace-aws-sdk-client

davidecarlier committed:
pass bucket to endpointProvider

davidecarlier commented on packages/drive-s3/src/AmazonWebServicesS3Storage.ts:
```diff
@@ -223,13 +250,13 @@
+		const endpoint = this.$driver.config.endpointProvider({ Region: this.$region });
 
-		if (href.startsWith('https://s3.amazonaws')) {
+		if (endpoint.url.href.startsWith('https://s3.amazonaws')) {
```
I removed this if statement in my latest commit and fixed the use of endpointProvider to generate the URL correctly.

rijkvanzanten reviewed (changes requested):
Seems like I'm getting an error when uploading files from Directus 🤔 

```
The \"path\" argument must be of type string. Received null

at AmazonWebServicesS3Storage._fullPath (/Users/rijkvanzanten/Developer/directus/directus/packages/drive-s3/dist/AmazonWebServicesS3Storage.js:53:61
```

davidecarlier committed:
fix endpoint & root config on s3client

davidecarlier commented:
@rijkvanzanten can you please try again now?

rijkvanzanten committed:
Merge branch 'main' into replace-aws-sdk-client

rijkvanzanten commented:
Different error now!

```
{
  "errors": [
    {
      "message": "Couldn't save file 9785ffee-5350-44ee-a6b1-2557e04809b5.jpg",
      "extensions": {
        "code": "SERVICE_UNAVAILABLE",
        "service": "files",
        "stack": "Error: Couldn't save file 9785ffee-5350-44ee-a6b1-2557e04809b5.jpg\n    at FilesService.uploadOne (/Users/rijkvanzanten/Developer/directus/directus/api/src/services/files.ts:77:10)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at Multipart.<anonymous> (/Users/rijkvanzanten/Developer/directus/directus/api/src/controllers/files.ts:90:23)"
      }
    }
  ]
}
```

from the API logs:

```
            Error code: TypeError
            Original stack:
            TypeError: Cannot read properties of null (reading 'length')
                at resolvedPath (/Users/rijkvanzanten/Developer/directus/directus/node_modules/.pnpm/@aws-sdk+smithy-client@3.201.0/node_modules/@aws-sdk/smithy-client/dist-cjs/resolve-path.js:8:24)
```

davidecarlier commented:
I can't reproduce the bug 😕 
Can you please share your configurations?

rijkvanzanten commented:
Using localhost minio as the S3 debugger (from the root folder of the repo `docker compose up -d minio`, with the following configuration:

```
STORAGE_LOCATIONS="minio"
STORAGE_MINIO_DRIVER="s3"
STORAGE_MINIO_KEY="minioadmin"
STORAGE_MINIO_SECRET="minioadmin"
STORAGE_MINIO_BUCKET="directus"
STORAGE_MINIO_ENDPOINT="localhost:5106"
```

Make sure to login to minio on localhost:5112 with minioadmin/minioadmin to create the `directus` bucket first

davidecarlier committed:
fix endpoint config

davidecarlier committed:
Merge branch 'main' into replace-aws-sdk-client

davidecarlier commented:
Hi @rijkvanzanten, I fixed a few things in the configuration of custom endpoints.
_On my machine_, it works both with a real s3 account and with minio


Side note, I tried to use minio on another directus instance (9.20.4) and it doesn't work:
`Inaccessible host: `directus.localhost' at port `undefined'`

So, I added directus.localhost to my hosts but now I get the error:
`MalformedXML: The XML you provided was not well-formed or did not validate against our published schema.`
Am I missing something about minio?

rijkvanzanten committed:
Merge branch 'main' into replace-aws-sdk-client
